keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+  from a node that is under maintenance.
+
+### Non-Goals
+- The PDB controller should detect and account for applications with evacuation capability when
+  calculating PDB status.
+- Introduce a field that could include non-critical daemon set pods
+  (priority `system-cluster-critical` or `system-node-critical`) for node maintenance/drain request.
+  The daemon set controller would then gracefully shut down these pods. Critical pods could be
+  overridden by the priority list mentioned below.
+- NodeMaintenance could include a plan of which pods to target first. Similar to graceful node
+  shutdown, we could include a list of priorities to decide which pods should be terminated first.
+  This list could optionally include pod timeouts, but could also wait for all the pods of a given
+  priority class to finish first without a timeout. This could also be used to target daemon set
+  pods of certain priorities (see point above). We could also introduce drain profiles based on
+  these lists. The cluster admin could then choose or create such a profile based on his/her needs.
+  The logic for processing the decision list would be contained in the node maintenance controller,
+  which would set an intent to selected pods to shut down via the EvacuationRequested
+  condition.
+- Introduce a node maintenance period, nodeDrainTimeout (similar to [cluster-api](https://cluster-api.sigs.k8s.io/developer/architecture/controllers/control-plane)
+  nodeDrainTimeout) or a TTL optional field as an upper bound on the duration of node maintenance.
+  Then the node maintenance would be garbage collected and the node made schedulable again.
+
+## Proposal
+Most of these issues stem from missing a standardized way of detecting a start of the node drain.
+This KEP proposes the introduction of a NodeMaintenance object that would signal an intent to
+gracefully remove pods from given nodes. The application pods should then signal back that the pods
+are being removed or migrated from the node. The implementation should also utilize existing node's
+`.spec.unschedulable` field, which prevents new pods from being scheduled on such a node.
+
+We will focus primarily on `kubectl drain` as a consumer of the NodeMaintenance API, but it can
+also be used by other drain implementations (e.g. node autoscalers) or manually. We will first
+introduce the API and then later modify the behavior of the Kubernetes system to fix all the node
+drain issues we mentioned earlier.
+
+To support workload migration, a new controller should be introduced to observe the NodeMaintenance
+objects and then mark pods for migration or removal with a condition. The pods would be selected
+according to the node at first (`nodeSelector`), but the selection mechanism can be extended later.
+Controllers can then implement the migration. The advantage of this approach is that controllers do
+not have to be aware of the NodeMaintenance object (no RBAC changes required). They only have to
```

Comment:
>The advantage of this approach is that controllers do
not have to be aware of the NodeMaintenance object (no RBAC changes required)

Can you please elaborate why the controllers won't be aware of the NodeMaintenance object? Shouldn't they act upon changes (creation/deletion/update) to this object? Should they be aware of node changes (maintenanced node's status is Not Ready)?

Reply:
Doing it this way means we can add the NodeMaintenance API without needing to change the whole ecosystem ahead of day 1 (which is handy!)

Reply:
I understand the consequences, but I am not sure how it will be implemented in a way to achieve this goal.
How can the NodeMaintenance controller would not be aware of the NodeMaintenance object and still be able to act upon its creation/update/deletion?

Reply:
Ah, try “controllers (apart from the NodeMaintenance controller)  do not have to be aware of the NodeMaintenance object kind”

The controller for NodeMaintenance does need to be coded; other controllers don't need updates.

Reply:
Oh yes, thanks for elaborating. It clarified my thoughts.

Reply:
The RBAC argument does not apply anymore. Currently some controllers need to be only aware of the Evacuation API (https://github.com/kubernetes/enhancements/pull/4565) and some of both the Evacuation API and the NodeMaintenance API (TBD the daemonset controller in this KEP)